main/Repl.hs

```diff
@@ -51,40 +46,135 @@ import           Control.Monad.Identity
+        banner
+        cmd
+        options
+        (Just commandPrefix)
+        (Just "paste")
+        completion
+        (rcFile >> greeter)
+        finalizer
+ where
+  commandPrefix = ':'
 
+  banner = pure . \case
+    System.Console.Repline.SingleLine -> "hnix> "
+    System.Console.Repline.MultiLine  -> "| "
 
-main :: (MonadNix e t f m, MonadIO m, MonadMask m) => m ()
-main = flip evalStateT initState
-    $ evalRepl (return prefix) cmd options (Just ':') completer welcomeText
- where
-  prefix = "hnix> "
-  welcomeText =
+  greeter =
     liftIO
       $  putStrLn
       $  "Welcome to hnix "
       <> showVersion version
       <> ". For help type :help\n"
+  finalizer = do
+    liftIO $ putStrLn "Goodbye."
+    return Exit
+
+  rcFile = do
+    f <- liftIO $ Data.Text.IO.readFile ".hnixrc" `catch` handleMissing
+    forM_ (map (words . Data.Text.unpack) $ Data.Text.lines f) $ \case
+      ((prefix:command) : xs) | prefix == commandPrefix -> do
+        let arguments = unwords xs
+        optMatcher command options arguments
+      x -> cmd $ unwords x
+
+  handleMissing e
+    | System.IO.Error.isDoesNotExistError e = return ""
+    | otherwise = throwIO e
+
+  -- Replicated and slightly adjusted `optMatcher` from `System.Console.Repline`
+  -- which doesn't export it.
+  -- * @MonadIO m@ instead of @MonadHaskeline m@
+  -- * @putStrLn@ instead of @outputStrLn@
+  optMatcher :: MonadIO m
+             => String
+             -> System.Console.Repline.Options m
+             -> String
+             -> m ()
+  optMatcher s [] _ = liftIO $ putStrLn $ "No such command :" ++ s
+  optMatcher s ((x, m) : xs) args
+    | s `Data.List.isPrefixOf` x = m args
+    | otherwise = optMatcher s xs args
 
 -------------------------------------------------------------------------------
 -- Types
 -------------------------------------------------------------------------------
 
-newtype IState t f m = IState
-  { tmctx :: AttrSet (NValue t f m)  -- Value environment
+data IState t f m = IState
+  { replIt  :: Maybe NExprLoc          -- ^ Last expression entered
+  , replCtx :: AttrSet (NValue t f m)  -- ^ Value environment
+  , replCfg :: ReplConfig              -- ^ REPL configuration
+  } deriving (Eq, Show)
+
+data ReplConfig = ReplConfig
+  { cfgDebug  :: Bool
+  , cfgStrict :: Bool
+  , cfgValues :: Bool
+  } deriving (Eq, Show)
+
+defReplConfig :: ReplConfig
+defReplConfig = ReplConfig
+  { cfgDebug  = False
+  , cfgStrict = False
+  , cfgValues = False
   }
 
-initState :: MonadIO m => IState t f m
-initState = IState M.empty
+-- | Create initial IState for REPL
+initState :: MonadNix e t f m => Maybe (NValue t f m) -> m (IState t f m)
+initState mIni = do
+
+  builtins <- evalText "builtins"
+
+  opts :: Nix.Options <- asks (view hasLens)
+
+  pure $ IState
+    Nothing
+    (Data.HashMap.Lazy.fromList $ Data.Maybe.catMaybes
+    [ (,) <$> pure "input" <*> mIni
+    , (,) <$> pure "builtins" <*> builtins
+    ])
```

Comment:
I think this could be simpler, especially when `evalText` doesn't return a `Maybe`. Maybe something like this?

```suggestion
    (Data.HashMap.Lazy.fromList $ ("builtins", builtins) : fmap ("input",) (Data.Maybe.maybeToList mIni))
```

Reply:
Thanks, TIL `TupleSections` & `maybeToList` :+1: